src/grpc/stream_infer_handler.cc

```diff
@@ -540,57 +544,104 @@ ModelStreamInferHandler::StreamInferResponseComplete(
-      const char* id;
-      LOG_TRITONSERVER_ERROR(
-          TRITONSERVER_InferenceResponseId(iresponse, &id),
-          "couldn't retrieve id for failed request");
-      LOG_VERBOSE(1) << "Failed for ID: " << id << std::endl;
-      response->mutable_infer_response()->set_id(id);
+      LOG_VERBOSE(1) << "Failed for ID: " << request_id << std::endl;
     }
 
     TRITONSERVER_ErrorDelete(err);
-
     LOG_TRITONSERVER_ERROR(
         TRITONSERVER_InferenceResponseDelete(iresponse),
         "deleting GRPC inference response");
   }
 
-  state->complete_ = ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) != 0);
-  if (!state->is_decoupled_) {
-    state->step_ = Steps::WRITEREADY;
-    state->context_->WriteResponseIfReady(state);
-  } else {
+  // Decoupled backends can return a null response via
+  // TRITONBACKEND_ResponseFactorySendFlags. By default, these null
+  // "flags_only" responses are not sent back to the client. Clients can
+  // opt-in to receiving these flags_only responses via request parameters.
+  // NOTE: The complete flag is the only flag used for this case at this time.
+  bool flags_only = (!iresponse && state->is_decoupled_ && state->complete_);
+  bool enable_flags_only = false;
+  if (flags_only) {
+    // Check request for opt-in parameter
+    const auto& request_params = state->request_.parameters();
+    const auto iter = request_params.find("triton_enable_flags_only_response");
+    if (iter != request_params.end()) {
+      const auto& param = iter->second;
+      if (param.parameter_choice_case() ==
+          inference::InferParameter::ParameterChoiceCase::kBoolParam) {
+        if (param.bool_param()) {
+          enable_flags_only = true;
+        }
+      } else {
+        LOG_ERROR << "invalid value type for "
+                     "'triton_enable_flags_only_response' parameter, expected "
+                     "bool_param.";
+      }
+    }
+  }
+
+  // Send flags_only response when detected and opted-in by client/request
+  const bool send_flags_only_response = (flags_only && enable_flags_only);
+  if (send_flags_only_response) {
+    // Create a response to send. Assume decoupled here based on prior checks.
+    state->response_queue_->AllocateResponse();
+    response = state->response_queue_->GetLastAllocatedResponse();
+    if (response) {
+      LOG_VERBOSE(1) << "[request id: " << request_id << "] "
+                     << "Sending flags-only response";
+      response->mutable_infer_response()->Clear();
+    } else {
+      LOG_ERROR << "expected the response allocator to have added the response";
+    }
+  }
+
+  // Set response parameters to indicate certain behaviors to client
+  if (response) {
+    response->mutable_infer_response()->set_id(request_id);
+    auto& params = *(response->mutable_infer_response()->mutable_parameters());
+    params["triton_final_response"].set_bool_param(state->complete_);
+    params["triton_flags_only_response"].set_bool_param(flags_only);
```

Comment:
Do we need to explicitly communicate to the client that this is a flags only response? Is there a way for the client to implicitly realize that it is a flags only response? For example, by not having any tensors in the output.

Reply:
This was brought up in the design doc, Guan had concerns with client differentiating between a response with no outputs/tensors being legitimate, or some hidden bug/error in server. Similarly, I don't think it's impossible for a model to have no outputs.

Do we have a contract/document somewhere saying that a model must have a non-zero number of outputs? I know it's certainly possible to return a response with no outputs because we don't verify it in core currently, but I don't know if it's "legal" in terms of what we expect.

Reply:
It is valid to have responses without any output tensor data. Some requests can just be intended to update the state of the model. We can have optional outputs in the responses.